ocaml/typing/typecore.ml

```diff
@@ -6900,21 +6949,40 @@ and type_cases
             (* allow propagation from preceding branches *)
             correct_levels ty_res
           else ty_res in
-        let guard =
+        let guard, exp =
           match pc_guard with
-          | None -> None
-          | Some scond -> (
-              match scond with
-              | Guard_predicate pred ->
-                Some
-                  (type_expect ext_env mode_local pred
-                    (mk_expected ~explanation:When_guard Predef.type_bool))
-              | Guard_pattern _ -> (failwith "pattern guard typechecking unimplemented")
-          )
-        in
-        let exp =
-          type_expect ?in_function ext_env emode
-            pc_rhs (mk_expected ?explanation ty_res')
+            | None | Some (Guard_predicate _) ->
+                let guard = match pc_guard with
+                  | None -> None
+                  | Some (Guard_predicate pred) ->
+                      let expected_bool =
+                        mk_expected ~explanation:When_guard Predef.type_bool
+                      in
+                      Some
+                        (Predicate
+                          (type_expect ext_env mode_local pred expected_bool))
+                  | Some (Guard_pattern _) -> assert false
+                in
+                let exp =
+                  type_expect
+                    ?in_function ext_env emode pc_rhs
+                    (mk_expected ?explanation ty_res')
+                in
+                guard, exp
+            | Some
+                (Guard_pattern 
+                  { pgp_scrutinee = e; pgp_pattern = pat; pgp_loc = loc }) ->
+                let { arg; sort; cases; partial = _ } =
```

Comment:
We will need this `partial` for the pattern match compiler &mdash; can you CR-soon? Resolve this upon addressing.